Revamp good practices (#10206)
pytest-dev/pytest: doc/en/explanation/goodpractices.rst

as well as the ``pytest`` package itself.
This ensures your code and dependencies are isolated from your system Python installation.

Create a ``pyproject.toml`` file in the root of your repository as described in
:doc:`packaging:tutorials/packaging-projects`.
The first few lines should look like this:

.. code-block:: toml

    [build-system]

>     requires = ["hatchling"]

Author: hmm, maybe we can also use tabs like in [the packaging tutorial][]?

I don’t recommend setuptools as it’s:

- slow
- messy (leaves a .egg-info folder lying around in your workspace, emits a bunch of unnecessary log messages during build)
- hard to configure (a bunch of legacy config files like MANIFEST.in, and misleadingly named config keys)
- no longer the first choice on [the packaging tutorial][]

but of course I can leave it if you prefer. I’d still get rid of `setup.cfg` since setuptools these days [supports PEP 621](https://setuptools.pypa.io/en/latest/userguide/quickstart.html). Unfortunately configuring it with more than just basic metadata is labelled as “beta”, e.g. this:

```toml
[tool.setuptools.packages.find]
where = ["src"]  # ["."] by default
```

[the packaging tutorial]: https://packaging.python.org/en/latest/tutorials/packaging-projects/

>     requires = ["hatchling"]

Reviewer: > no longer the first choice on [the packaging tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/)

Ahh thanks for pointing that out. If the packaging tutorial recommends `hatchling` as first choice, I think we should follow suit then.

>     requires = ["hatchling"]

Author: In case you would like some context, I followed that discussion, and the reasoning went something like that:

1. we don’t want to promote any standards following build backend
2. so we use tabs
3. but since the first (auto-selected) tab is kind of an endorsement, we need something there that is good for the vast majority of people, i.e. pure python packages
4. because of the above mentioned drawbacks of setuptools, the first tab will be hatchling, because it provides a better experience for people creating a pure python package than setuptools

Hatchling doesn’t have a sufficient amount of plugins or example code yet to allow a highly custom build like numpy or numba or so, but for anything with purely static build configuration, it’s a very good choice.

    See :ref:`pytest vs python -m pytest` for more information about the difference between calling ``pytest`` and
    ``python -m pytest``.

> For the most convenient experience,
> choose the ``import-lib`` :ref:`import mode <import-modes>`.
> To this end, add the following to your ``pyproject.toml``:

Reviewer: ```suggestion
For new projects, we recommend to use ``importlib`` :ref:`import mode <import-modes>` (see [1]_ for a detailed explanation).
To this end, add the following to your ``pyproject.toml``:
```

.. code-block:: toml

    [tool.pytest.ini_options]
    addopts = [
        "--import-mode=importlib",
    ]

> The default :ref:`import mode <import-modes>` ``prepend`` has several drawbacks [1]_.

Author: Hmm, my understanding is that there‘s only the property that you can’t import from test files, which is an advantage in my book. It’s only a blocker for adoption by existing projects because it’s not backwards compatible, which I should mention.

Are there any disadvantages apart from “isn’t backwards compatible in this regard”?

> The default :ref:`import mode <import-modes>` ``prepend`` has several drawbacks [1]_.

Reviewer: @flying-sheep gentle ping here just in case you missed this comment. 👍

> The default :ref:`import mode <import-modes>` ``prepend`` has several drawbacks [1]_.

Author: Sure, I’ll spell out the differences there! Due to the amount of misunderstandings of the good practices document in the past, I think there should be several references, so I think referring to that footnote/section here in some form isn’t the worst idea.

.. _`src-layout`:

> Generally, but especially if you use the default import mode,

Reviewer: ```suggestion
Generally, but especially if you use the default import mode ``prepend``,
```

> Generally, but especially if you use the default import mode,

Reviewer: I agree we should primarily recommend the src code layout, so it makes sense to move this layout to the top of the session.

Author: will do!

it is **strongly** suggested to use a ``src`` layout.
Here, your application root package resides in a sub-directory of your root:

.. code-block:: text

> 

Reviewer: (I can't comment on it because it is not part of the diff)

We should remove `setup.cfg` from the listing below given we did the same with the first listing.

            __init__.py
            app.py
            view.py
    tests/...

This layout prevents a lot of common pitfalls and has many benefits, which are better explained in this excellent
`blog post by Ionel Cristian Mărieș <https://blog.ionelmc.ro/2014/05/25/python-packaging/#the-structure>`_.

> Tests as part of application code

Reviewer: We should remove `setup.cfg` from the listing below too (I can't comment directly on it because it is not part of the diff).

setuptools intends to
`remove the test command <https://github.com/pypa/setuptools/issues/931>`_.

.. [1] The default ``prepend`` :ref:`import mode <import-modes>` works as follows:

    Since there are no packages to derive a full package name from,
    ``pytest`` will import your test files as *top-level* modules.
    The test files in the first example would be imported as
    ``test_app`` and ``test_view`` top-level modules by adding ``tests/`` to ``sys.path``.

>     This results in a drawback compared to the import mode ``import-lib``:

Reviewer: ```suggestion
    This results in a drawback compared to the import mode ``importlib``:
```
